lib/heuristics/dichotomious_approach.rb

```diff
@@ -149,13 +144,22 @@ def self.transfer_unused_vehicles(result, sub_service_vrps)
 
         sv_one.vehicles << vehicle
         sv_zero.vehicles -= [vehicle]
-        sv_one.points += sv_zero.points.select{ |p| p.id == vehicle.start_point_id || p.id == vehicle.end_point_id }
+        sv_one.points += [vehicle.start_point, vehicle.end_point].compact
       }
 
       # Transfer unsued vehicle limit to the other side as well
       sv_zero_used_vehicle_count = result[:routes].count{ |r| r[:activities].any?{ |a| a[:service_id] } }
       sv_zero_unused_vehicle_limit = sv_zero.resolution_vehicle_limit - sv_zero_used_vehicle_count
       sv_one.resolution_vehicle_limit += sv_zero_unused_vehicle_limit
+
+      return if sv_zero_used_vehicle_count == 0
+
+      # Transfered vehicles requires to update the current matrix
+      matrix_indices = sv_one.points.map{ |point|
+        Models::Point.find_by_id(point.id).matrix_index
```

Comment:
this should work, right? Is there a specific reason to go via  Models::Point?
```suggestion
        point.matrix_index
```